### PR TITLE
Adding design doc for multi device(background)

### PR DESCRIPTION
**Pull request by kavyasrinet**
This is the design document that explains the background of multi-device training.
The next steps are:
1. Design doc for how multi device happens in Tensorflow
2. Design doc for our Python API.
3. Design doc for ProgramDesc

### CODE REVIEW

File: doc/design/multi_device_training.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# Multi-device training in PaddlePaddle
+
+## Why Multi-device
+On a typical device or system, there are multiple computing devices. In PaddlePaddle, the supported device types as of now, are CPU and GPU (we will be adding FPGA in the future too).
+
+If a PaddlePaddle operation has both CPU and GPU implementations, we decide which kernel to execute based on the device type.  
+Training deep learning models can be resource intensive. Even with a very powerful GPU, some models can take really long to train. This is obvious with deep learning models, especially recurrent models where the execution of each step depends on the execution and output of previous step.
```
Comment (QiJune) [inline doc/design/multi_device_training.md]: Training is a scene of using multi-device. In inference, we may also need to support multi-device. For example, FPGA is used for inference, but FPGA is not suitable for all operators. We may switch to CPU for some complex operators and then, switch back to FPGA.

Comment (QiJune) [reply doc/design/multi_device_training.md]: @QingshuChen and @shijiaxin Please kindly review this doc. We are planning to support multi-device feature for Paddle.

**Comment by QiJune** [comment]
@QingshuChen and @shijiaxin Please kindly review this doc. We are planning to support multi-device feature for Paddle.

File: doc/design/multi_device_training.md (hunk 1 above)
Comment (kavyasrinet) [reply doc/design/multi_device_training.md]: Thanks for pointing this out. Will update the writeup.

File: doc/design/multi_device_training.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+1. Data Parallelism
+2. Model Parallelism
+
+### Data Parallelism
+Data parallelism works by partitioning the training data over all the devices and hence distributes the workload over multiple devices. Each device has a copy of the complete model and only has access to 1/d of the total training data (if there are d devices in total). The updates from each device (gradients, parameter updates etc.) are communicated across all the devices once the device has an update.
+
+### Model Parallelism
+Model parallelism on the other hand, works by keeping a part of the model on each available device. This is useful when the model is too large to keep on one device or when there are parts of the model that can be executed independently ( in parallel). In this setup, each device will train a part of the model and pass on its updates to the next device.
+
+Here, we want to explore the model parallelism setup, where different parts of the same model reside on different devices and communicate to each other by sending updates.
```
[the remark this answers is not in the document]
Comment (kavyasrinet) [reply doc/design/multi_device_training.md]: As per my discussion with @QiJune , he mentioned that we should focus on model parallelism at the moment, some other team is already looking at data parallelism. If that's not the case, I would be happy to include more about data parallelism as well.

[the remark this answers is not in the document]
Comment (zealoct) [reply doc/design/multi_device_training.md]: Hello, I think maybe computation node(operator) placement policy is another topic that is worth mentioning in multi-device situation. 
Sometimes the model is not parallel, but it still benefit the overall training and inference efficiency to put some compute nodes on devices like GPU or FPGA, like DNN/CNN part. Of course user could be able to explicitly specify on which device each of the model's node is computed, but I think how to assign each node's computing device automatically is worth discussing.

File: doc/design/multi_device_training.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+We have some low level implementations of CUDA memcpy [here](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/memory/memcpy.h) and the next step would be to write C++ operators to expose these functions.
+
+### Python API
+To enable users to use the setup we talked about above, we need to set up a Python API that the users can use. We specifically need two components here:
+1. The Python API: This is the interface that the users of PaddlePaddle will use to set up model/data parallelism.
+2. ProgramDesc: Behind the scenes we need a module that can convert the python API to a ProgramDesc (which is a collection of repeated OpDescs). The ProgramDesc then will be sent to the Executor, which creates the Ops and eventually runs the Ops.
+
+We need to design the above two components as well as propose how the Python API will be parsed into ProgramDesc.
+These components will be addressed in the following design documents, one for each component.
```
[the remark this answers is not in the document]
Comment (kavyasrinet) [reply doc/design/multi_device_training.md]: Sounds good. Fixed in the recent commit.

File: doc/design/multi_device_training.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (kavyasrinet) [reply doc/design/multi_device_training.md]: Thanks for pointing that out, I covered how TensorFlow does that in this PR: https://github.com/PaddlePaddle/Paddle/pull/5412 , but this is a good suggestion, I will fix this PR to include this too.

[the remark this answers is not in the document]
Comment (kavyasrinet) [reply doc/design/multi_device_training.md]: Hello @zealoct , I have fixed the changes you proposed in the latest commit. Please have a look when you can.

**Pull request closed without merging**